client/client.go

```diff
@@ -353,94 +354,75 @@ func (kc *kClient) applyCallOptions(ctx context.Context, cfg rpcinfo.MutableRPCC
+			rpcinfo.PutRPCInfo(ri)
+		}
+		callOpts.Recycle()
+	}()
+
+	callOptRetry, callOptFallback := initRetryAndFallbackWithCallopts(callOpts)
 	if kc.opt.RetryContainer == nil && callOptRetry.Enable {
 		// setup retry in callopt
 		kc.opt.RetryContainer = retry.NewRetryContainer()
 	}
 
-	var err error
-	var recycleRI bool
 	if kc.opt.RetryContainer == nil {
+		// call without retry policy
 		err = kc.eps(ctx, request, response)
 		if err == nil {
 			recycleRI = true
 		}
 	} else {
-		var callTimes int32
-		// prevRI represents a value of rpcinfo.RPCInfo type.
-		var prevRI atomic.Value
-		recycleRI, err = kc.opt.RetryContainer.WithRetryIfNeeded(ctx, callOptRetry, func(ctx context.Context, r retry.Retryer) (rpcinfo.RPCInfo, interface{}, error) {
-			currCallTimes := int(atomic.AddInt32(&callTimes, 1))
-			retryCtx := ctx
-			cRI := ri
-			if currCallTimes > 1 {
-				retryCtx, cRI, _ = kc.initRPCInfo(ctx, method)
-				retryCtx = metainfo.WithPersistentValue(retryCtx, retry.TransitKey, strconv.Itoa(currCallTimes-1))
-				if prevRI.Load() == nil {
-					prevRI.Store(ri)
-				}
-				r.Prepare(retryCtx, prevRI.Load().(rpcinfo.RPCInfo), cRI)
-				prevRI.Store(cRI)
-			}
-			err := kc.eps(retryCtx, request, response)
-			return cRI, response, err
-		}, ri, request)
+		recycleRI, err = kc.opt.RetryContainer.WithRetryIfNeeded(ctx, callOptRetry, kc.rpcCallWithRetry(ri, method, request, response), ri, request)
 	}
 
 	// do fallback if with setup
-	fallback, hasFallback := getFallbackPolicy(callOptFallback, kc.opt.Fallback)
-	var fbErr error
-	reportErr := err
-	if hasFallback {
-		reportAsFB := false
-		// Notice: If rpc err is nil, rpcStatAsFB will always be false, even if it's set to true by user.
-		fbErr, reportAsFB = fallback.DoIfNeeded(ctx, ri, request, response, err)
-		if reportAsFB {
-			reportErr = fbErr
-		}
-		err = fbErr
-	}
-
-	kc.opt.TracerCtl.DoFinish(ctx, ri, reportErr)
-	callOpts.Recycle()
-	if err == nil && !hasFallback {
-		err = ri.Invocation().BizStatusErr()
-	}
+	err, reportErr = doFallbackIfNeeded(ctx, ri, request, response, err, callOptFallback, kc.opt.Fallback)
+	return err
+}
 
-	if recycleRI {
-		// why need check recycleRI to decide if recycle RPCInfo?
-		// 1. no retry, rpc timeout happen will cause panic when response return
-		// 2. retry success, will cause panic when first call return
-		// 3. backup request may cause panic, cannot recycle first RPCInfo
-		// RPCInfo will be recycled after rpc is finished,
-		// holding RPCInfo in a new goroutine is forbidden.
-		rpcinfo.PutRPCInfo(ri)
+func (kc *kClient) rpcCallWithRetry(ri rpcinfo.RPCInfo, method string, request, response interface{}) retry.RPCCallFunc {
+	// call with retry policy
+	var callTimes int32
+	// prevRI represents a value of rpcinfo.RPCInfo type.
+	var prevRI atomic.Value
+	return func(ctx context.Context, r retry.Retryer) (rpcinfo.RPCInfo, interface{}, error) {
+		currCallTimes := int(atomic.AddInt32(&callTimes, 1))
+		retryCtx := ctx
```

Comment:
Why the need to copy ctx here? It's already copied when calling this function.

Reply:
done